docs/design/proposals/cephfs-snapshot-shallow-ro-vol.md

```diff
@@ -0,0 +1,274 @@
+  because it still has snapshots associated to it.
+* _Source snapshot from which the shallow volume originates is removed while
+  that shallow volume still exists?_
+
+  We need to make sure this doesn't happen and some book-keeping
+  is necessary. Ideally we could employ some kind of reference counting.
+
+#### Reference counting for shallow volumes
+
+As mentioned above, this is to protect shallow volumes, should their source
+snapshot be requested for deletion.
+
+When creating a volume snapshot, a new RADOS object would be created for that
+snapshot. It would contain a single signed integer value that would describe
+the number of references the backing subvolume snapshot has. It would be
+initialized to one, where the CSI snapshot itself is the first reference to the
+backing snapshot. Each subsequent shallow volume creation would increment the
+reference counter, and each shallow volume deletion would decrement it.
+Calling `DeleteSnapshot` would decrement the reference counter as well.
+
+The subvolume snapshot would be removed from the Ceph cluster only once the
+reference count reaches zero. Note that this behavior would permit calling
+`DeleteSnapshot` even if it is still referenced by shallow volumes.
+
+* `DeleteSnapshot`:
+  * `refcount == 0` or the respective _refcount_ RADOS object doesn't exist:
+    delete the backing snapshot too.
+  * `refcount > 0`: keep the backing snapshot.
+* `DeleteVolume`:
+  * `refcount == 0`: delete the backing snapshot too.
+  * `refcount > 0`: keep the backing snapshot.
+
+To enable creating shallow volumes from snapshots that were provisioned by
+older versions of cephfs-csi (i.e. before this feature is introduced),
+`CreateVolume` for shallow volumes would also create the _refcount_ object in
+case it's missing. It would be initialized to two: the source snapshot and the
+newly created shallow volume.
+
+##### Concurrent access to _refcount_ objects
+
+We can't do operations on integers in RADOS objects atomically, which means we
+need to do all synchronization ourselves when modifying the value of the counter.
+
+From a single master perspective, concurrent access to _refcount_ can be
+protected by mutex locks and is easy to implement. `cephfs.ControllerServer`
+already offers `VolumeLocks`, `SnapshotLocks` and `OperationLocks`.
+
+Mutexes won't cover situations where multiple masters may be active though
+(during/after leader election), and modifying _refcount_ would still race in
+this case. Misbehaving master that lost leadership would still be able to
+access resources on the Ceph cluster. Racing on _refcount_ could then cause
+leaking the _refcount_ object itself and/or the subvolume snapshot, or the
+provisioned shallow volume could point to a subvolume snapshot that's already
+been deleted. This is not easily fixable without the individual master replicas
+being aware of the leader election process and is out-of-scope for this
+proposal.
+
```

Comment:
Yeah I was thinking about this scenario too. I see that in another commet you already suggested setting Deployment replicas to 1. I'm wondering whether it shouldn't be the default anyway though, and whether it's enough to protect from similar races in general. With Deployments r=1 there still could be more than one replica running concurrently (albeit the others would be in Terminating state), so there's still a chance for a race to occur no? StatefulSets provide the uniqueness guarantee which would take care of this problem altogether.

Whenever ceph-csi needs a mutex to protect Ceph resources from concurrent access, it's usually a good indicator some controller locking / fencing is also necessary, should a leader switch occur. What leader election provides in sidecars is simply not enough:

https://github.com/kubernetes/client-go/blob/release-14.0/tools/leaderelection/leaderelection.go#L17-L20

How much availability would ceph-csi lose if it were to simply ditch leader election and use a single-replica StatefulSets instead? Please correct me if I'm wrong, but isn't the behavior for resolving failure with and without leader election very similar? 
* With leader election: leader controller plugin crashes / becomes unresponsive, another replica is selected as the new leader.
* Without leader election: controller plugin crashes / becomes unresponsive (signaled by a liveness probe), Kubernetes takes care of recreating the Pod.

In both of those cases some unavailability is present whether it is due to timing out and finding a new leader, or creating a new Pod. Except that multiple replicas + leader election run the risk of race conditions. Maybe we can discuss this more in a separate ticket.

Reply:
Yes it's a bit tricky and needs a lot of testing. we moved from statefulset to deployment due to the downtime issues statefulset. and deployment with replica:2/3 got added to have the HA. but with this one, we have a leader election problem. @ShyamsundarR opened an issue for this one https://github.com/ceph/ceph-csi/issues/742 I think we need to check with the CSI team as this can be a problem? but if CSI is stateless this does not matter but for us, it matters as things can go wrong with the counter.

Reply:
Thanks for the pointer, @Madhu-1 ! I suppose this is the problem you've mentioned https://github.com/ceph/ceph-csi/pull/414 ?

Reply:
Yes, correct that's the one.